### PR TITLE
process_sstables_dir: close directory_lister on error

### LINKED ISSUE #10697: Unrecognized files in table data directory triggers assert failure
Putting any file unrecognizable to scylla as a valid sstable component will trigger an assert failure as follows:
```
INFO  2022-05-30 17:36:23,088 [shard 0] database - Populating Keyspace ks1                                                                                                                                                                                    
INFO  2022-05-30 17:36:23,088 [shard 0] database - Keyspace ks1: Reading CF table1 id=0a75f5b0-e024-11ec-96c6-c7dc2b0ff2ed version=21acc750-47d1-3167-a77d-ec3c544d1525                                                                                       terminate called after throwing an instance of 'seastar::internal::backtraced<std::runtime_error>'                                                                                                                                                            
  what():  directory_lister not closed when destroyed Backtrace: 0x3443cce 0x34441c0 0x34444c8 0x120c902 0x32f08f3 0x32f0794 0x32f0604 0x20b2a05 0x153b40e 0x3315ba4 0x3316e97 0x33161c9 0x32cdd4f 0x32cd23e 0xf1eb7e 0xf1dae0 /lib64/libc.so.6+0x4043f /lib64
/libc.so.6+0x404ef 0xf1d3e4
```
Reproducer: creating an empty `out.json` file in `/var/lib/scylla/data/system/large_rows-*/`.

@bhalevy

### PR DESCRIPTION
**Pull request by bhalevy**
`sstable_directory::process_sstables_dir` may hit an exception when calling `handle_component`.
In this case we currently destroy the `sstable_dir_lister` variable without closing the `directory_lister` first -
leading to terminate in `~directory_lister` as seen in #10697.

This mini-series handles this exception and always closes the `directory_lister`.
    
Add unit test to reproduce this issue.
    
Fixes #10697

### CODE REVIEW

**Comment by bhalevy** [comment]
Converted to draft since this depends on https://github.com/scylladb/seastar/pull/1090
to properly propagate the exception type via `co_await coroutine::return_exception(std::move(ex))`

**Review by denesb — APPROVED** [review]

**Comment by bhalevy** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/740/

https://jenkins.scylladb.com/job/releng/job/Scylla-CI/740/artifact/testlog/x86_64_release/sstable_directory_test.sstable_directory_test_table_scan_invalid_file.1657.log
```
unknown location(0): fatal error: in "sstable_directory_test_table_scan_invalid_file": unknown type
```

is fixed by https://github.com/scylladb/seastar/pull/1090

**Comment by bhalevy** [comment]
In v2 (89c5e8413f)
- rethrow caught exception_ptr rather than using `co_await coroutine::return_exception` until https://github.com/scylladb/seastar/pull/1090 is merged and gets to scylla.

**Comment by denesb** [comment]
Will merge once CI approves.

**Pull request merged**
